src/librbd/api/Config.cc

```diff
@@ -22,85 +23,51 @@ const uint32_t MAX_KEYS = 64;
-      {"rbd_journal_object_flush_age", {}},
-      {"rbd_journal_object_flush_bytes", {}},
-      {"rbd_journal_object_flush_interval", {}},
-      {"rbd_journal_order", {}},
-      {"rbd_journal_pool", {}},
-      {"rbd_journal_splay_width", {}},
-      {"rbd_localize_parent_reads", {}},
-      {"rbd_localize_snap_reads", {}},
-      {"rbd_mirroring_delete_delay", {}},
-      {"rbd_mirroring_replay_delay", {}},
-      {"rbd_mirroring_resync_after_disconnect", {}},
-      {"rbd_mtime_update_interval", {}},
-      {"rbd_non_blocking_aio", {}},
-      {"rbd_qos_bps_limit", {}},
-      {"rbd_qos_iops_limit", {}},
-      {"rbd_qos_read_bps_limit", {}},
-      {"rbd_qos_read_iops_limit", {}},
-      {"rbd_qos_write_bps_limit", {}},
-      {"rbd_qos_write_iops_limit", {}},
-      {"rbd_readahead_disable_after_bytes", {}},
-      {"rbd_readahead_max_bytes", {}},
-      {"rbd_readahead_trigger_requests", {}},
-      {"rbd_request_timed_out_seconds", {}},
-      {"rbd_skip_partial_discard", {}},
-      {"rbd_sparse_read_threshold_bytes", {}},
-      // rbd-mirror daemon options
-      {"rbd_mirror_concurrent_image_deletions", {}},
-      {"rbd_mirror_concurrent_image_syncs", {}},
-      {"rbd_mirror_delete_retry_interval", {}},
-      {"rbd_mirror_image_policy_migration_throttle", {}},
-      {"rbd_mirror_image_policy_rebalance_timeout", {}},
-      {"rbd_mirror_image_policy_type", {}},
-      {"rbd_mirror_image_policy_update_throttle_interval", {}},
-      {"rbd_mirror_image_state_check_interval", {}},
-      {"rbd_mirror_journal_commit_age", {}},
-      {"rbd_mirror_journal_max_fetch_bytes", {}},
-      {"rbd_mirror_journal_poll_age", {}},
-      {"rbd_mirror_leader_heartbeat_interval", {}},
-      {"rbd_mirror_leader_max_acquire_attempts_before_break", {}},
-      {"rbd_mirror_leader_max_missed_heartbeats", {}},
-      {"rbd_mirror_sync_point_update_age", {}},
-    } {
-    if (!image_apply_only_options) {
-      Parent image_create_opts = {
-        {"rbd_default_data_pool", {}},
-        {"rbd_default_features", {}},
-        {"rbd_default_order", {}},
-        {"rbd_default_stripe_count", {}},
-        {"rbd_default_stripe_unit", {}},
-        {"rbd_journal_order", {}},
-        {"rbd_journal_pool", {}},
-        {"rbd_journal_splay_width", {}},
-      };
-      insert(image_create_opts.begin(), image_create_opts.end());
+  librados::IoCtx& io_ctx;
+
+  Options(librados::IoCtx& io_ctx, bool image_apply_only_options)
+    : io_ctx(io_ctx) {
+    CephContext *cct = reinterpret_cast<CephContext *>(io_ctx.cct());
+
+    const std::string rbd_key_prefix("rbd_");
+    auto& schema = cct->_conf.get_schema();
+    for (auto& pair : schema) {
+      if (!boost::starts_with(pair.first, rbd_key_prefix)) {
+        continue;
+      } else if (EXCLUDE_OPTIONS.count(pair.first) != 0) {
+        continue;
+      } else if (image_apply_only_options &&
+                 EXCLUDE_IMAGE_OPTIONS.count(pair.first) != 0) {
+        continue;
+      }
```

Comment:
@dillaman Now `rbd config pool/image list` returns much more options. I have reviewed the list and yes there are options that were erroneously not listed previously, but there are those that are not applicable on pool/image level or currently erroneously not applicable. 

For pool level the list:
- rbd_blkin_trace_all (applicable, it was a bug that the earlier version of "config list" ignored it)
- rbd_default_clone_format (bug: librbd/image/CloneRequest.cc needs fixing to apply a pool level override)
- rbd_default_map_options (bug: tools/rbd/action/Kernel.cc needs fixing to apply a pool level override)
- rbd_default_format (bug: librbd/internal.cc needs fixing to apply a pool level override)
- rbd_default_pool (can't be applied on pool level)
- rbd_discard_on_zeroed_write_same (bug: src/librbd/librbd.cc is fixed to apply pool/image level override but src/tools/rbd/action/Create.cc needs fixing to ignore these overrides)
- rbd_enable_alloc_hint (applicable, it was a bug that the earlier version of "config list" ignored it)
- rbd_journal_object_max_in_flight_appends (applicable, it was a bug that earlier version of "config list" ignored it)
- rbd_mirror_pool_replayers_refresh_interval (can't be applied on pool level)
- rbd_tracing (can't be applied on pool level)

For image level the same issues that are for pool level + listed "rbd_mirror_" options that are not applicable here.

Reply:
@trociny Thx -- updated.